### PR TITLE
Implement Big UInt Bitwise Or for modexp

### LINKED ISSUE #132: Implement Big UInt Bitwise Or for `modexp`
**Context:** `modexp.yul`

**Description:** 

```solidity
/// @notice Performs the big unsigned integer bit or operation.
/// @dev The result is stored from `resPtr` to `resPtr + (WORD_SIZE * nLimbs)`.
bigUIntBitOr(lhsPtr, rhsPtr, nLimbs, resPtr)
```

`lhsPtr`: The pointer to the MSB of the left operand.
`rhsPtr`: The pointer to the MSB of the right operand.
`nLimbs`: The number of limbs needed to represent the operands.
`resPtr`: The pointer to where you want the result to be stored.

**Recommendation:**

Follow [this implementation](https://github.com/lambdaclass/lambdaworks/blob/main/math/src/unsigned_integer/element.rs#L346).

### PR DESCRIPTION
**Pull request by jpcenteno**
closes #132

### CODE REVIEW

File: precompiles/Modexp.yul [hunk 1]
```diff
@@ -67,6 +67,22 @@ object "ModExp" {
...
+            /// @param rhsPtr The pointer to the MSB of the right operand.
+            /// @param nLimbs The number of limbs needed to represent the operands.
+            /// @param resPtr The pointer to where you want the result to be stored
+            function bigUIntBitOr(lhsPtr, rhsPtr, nLimbs, resPtr) {
+                for { let i := 0 } lt(i, nLimbs) { i := add(i, 1) }
+                {
+                    lhs_i := mload(add(lhsPtr, i))
+                    rhs_i := mload(add(rhsPtr, i))
+                    res_i := or(lhs_i, rhs_i)
+                    mstore(add(resPtr, i), res_i)
```
Comment (ilitteri) [inline precompiles/Modexp.yul:82]: We need to add 32 per limb to move to the next memory slot:
```suggestion
                    let offset_i := mul(WORD_SIZE(), i)
                    lhs_i := mload(add(lhsPtr, offset_i))
                    rhs_i := mload(add(rhsPtr, offset_i))
                    res_i := or(lhs_i, rhs_i)
                    mstore(add(resPtr, offset_i), res_i)
```

File: precompiles/Modexp.yul [hunk 2]
```diff
@@ -67,6 +67,22 @@ object "ModExp" {
...
+            /// @param lhsPtr The pointer to the MSB of the left operand.
+            /// @param rhsPtr The pointer to the MSB of the right operand.
+            /// @param nLimbs The number of limbs needed to represent the operands.
+            /// @param resPtr The pointer to where you want the result to be stored
+            function bigUIntBitOr(lhsPtr, rhsPtr, nLimbs, resPtr) {
+                for { let i := 0 } lt(i, nLimbs) { i := add(i, 1) }
+                {
+                    lhs_i := mload(add(lhsPtr, i))
+                    rhs_i := mload(add(rhsPtr, i))
+                    res_i := or(lhs_i, rhs_i)
```
Comment (ilitteri) [inline precompiles/Modexp.yul:81]:
```suggestion
                    let lhs_i := mload(add(lhsPtr, i))
                    let rhs_i := mload(add(rhsPtr, i))
                    let res_i := or(lhs_i, rhs_i)
```

**Review by ilitteri — CHANGES_REQUESTED** [review]
Looking good! Just some minor details

File: precompiles/Modexp.yul [hunk 3]
```diff
@@ -67,6 +67,37 @@ object "ModExp" {
...
+                // |            |                       |                               |                               |                               |                 |                 |                                      |
+                // | nLimbs - 1 | +(0x20 * (nLimbs - 1) | lhsPtr + (0x20 * (nLimbs - 1) | rhsPtr + (0x20 * (nLimbs - 1) | resPtr + (0x20 * (nLimbs - 1) | lhs[nLimbs - 1] | rhs[nLimbs - 1] | or(lhs[nLimbs - 1], rhs[nLimbs - 1]) |
+                // +------------+-----------------------+-------------------------------+-------------------------------+-------------------------------+-----------------+-----------------+--------------------------------------+
+
+                let finalOffset := shl(5, nLimbs) // == ( LIMB_SIZE * nLimbs ) == (32 * nLimbs) 
+                for { let offset_i := 0 } lt(offset_i, finalOffset) { offset_i := add(offset_i, 0x20) }
+                {
+                    let ptr_lhs_i := add(lhsPtr, offset_i)
+                    let ptr_lhs_i := add(rhsPtr, offset_i)
```
Comment (ilitteri) [inline precompiles/Modexp.yul:93]:
```suggestion
                    let ptr_rhs_i := add(rhsPtr, offset_i)
```

**Review by ilitteri — CHANGES_REQUESTED** [review]
I think one variable name is wrong.

File: precompiles/Modexp.yul [hunk 4]
```diff
@@ -67,6 +67,37 @@ object "ModExp" {
...
+                    let ptr_res_i := add(resPtr, offset_i)
+                    let value_lhs_i := mload(ptr_lhs_i)
+                    let value_rhs_i := mload(ptr_lhs_i)
+                    let value_res_i := or(value_lhs_i, value_rhs_i)
```
[the remark this answers is not in the document]
Comment (jpcenteno) [reply precompiles/Modexp.yul]: TY! I might have broke this when copying and pasting, IDK.

**Review by ilitteri — APPROVED** [review]
LGTM!

## How to test

1. Copy the function and its dependencies into a `Playground.yul` file.
2. Copy the following fallback in the file.
```solidity
// BIG UINT OR TESTS

// 1 Limb
let lhsPtr := 0x00
let rhsPtr := 0x20
let orPtr := 0x40
let nLimbs := 1

// 0000 | 0000 = 0000
mstore(lhsPtr, 0x00)
mstore(rhsPtr, 0x00)
bigUIntBitOr(lhsPtr, rhsPtr, nLimbs, orPtr)
console_log(mload(orPtr)) // 0x00

// 1111 | 0000 = 1111
mstore(lhsPtr, 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
mstore(rhsPtr, 0x00)
bigUIntBitOr(lhsPtr, rhsPtr, nLimbs, orPtr)
console_log(mload(orPtr)) // 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff

// 0000 | 1111 = 1111
mstore(lhsPtr, 0x00)
mstore(rhsPtr, 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
bigUIntBitOr(lhsPtr, rhsPtr, nLimbs, orPtr)
console_log(mload(orPtr)) // 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff

// 1111 | 1111 = 1111
mstore(lhsPtr, 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
mstore(rhsPtr, 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
bigUIntBitOr(lhsPtr, rhsPtr, nLimbs, orPtr)
console_log(mload(orPtr)) // 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff

// 1010 | 0101 = 1111
mstore(lhsPtr, 0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa)
mstore(rhsPtr, 0x5555555555555555555555555555555555555555555555555555555555555555)
bigUIntBitOr(lhsPtr, rhsPtr, nLimbs, orPtr)
console_log(mload(orPtr)) // 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff

// 0101 | 1010 = 1111 (needed?)
mstore(lhsPtr, 0x5555555555555555555555555555555555555555555555555555555555555555)
mstore(rhsPtr, 0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa)
bigUIntBitOr(lhsPtr, rhsPtr, nLimbs, orPtr)
console_log(mload(orPtr)) // 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff

// 2 Limbs
let lhsPtr := 0x00
let rhsPtr := 0x40
let orPtr := 0x80
let nLimbs := 2

// 0000 0000 | 0000 0000 = 0000 0000
mstore(lhsPtr, 0x00)
mstore(add(lhsPtr, 0x20), 0x00)
mstore(rhsPtr, 0x00)
mstore(add(rhsPtr, 0x20), 0x00)
bigUIntBitOr(lhsPtr, rhsPtr, nLimbs, orPtr)
console_log(mload(orPtr)) // 0x00
console_log(mload(add(orPtr, 0x20))) // 0x00

// 1111 1111 | 0000 0000 = 1111 1111
mstore(lhsPtr, 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
mstore(add(lhsPtr, 0x20), 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
mstore(rhsPtr, 0x00)
mstore(add(rhsPtr, 0x20), 0x00)
bigUIntBitOr(lhsPtr, rhsPtr, nLimbs, orPtr)
console_log(mload(orPtr)) // 0xff
console_log(mload(add(orPtr, 0x20))) // 0xff

// 0000 0000 | 1111 1111 = 1111 1111
mstore(lhsPtr, 0x00)
mstore(add(lhsPtr, 0x20), 0x00)
mstore(rhsPtr, 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
mstore(add(rhsPtr, 0x20), 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
bigUIntBitOr(lhsPtr, rhsPtr, nLimbs, orPtr)
console_log(mload(orPtr)) // 0xff
console_log(mload(add(orPtr, 0x20))) // 0xff

// 1111 1111 | 1111 1111 = 1111 1111
mstore(lhsPtr, 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
mstore(add(lhsPtr, 0x20), 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
mstore(rhsPtr, 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
mstore(add(rhsPtr, 0x20), 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
bigUIntBitOr(lhsPtr, rhsPtr, nLimbs, orPtr)
console_log(mload(orPtr)) // 0xff
console_log(mload(add(orPtr, 0x20))) // 0xff

// 1010 1010 | 0101 0101 = 1111 1111
mstore(lhsPtr, 0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa)
mstore(add(lhsPtr, 0x20), 0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa)
mstore(rhsPtr, 0x5555555555555555555555555555555555555555555555555555555555555555)
mstore(add(rhsPtr, 0x20), 0x5555555555555555555555555555555555555555555555555555555555555555)
bigUIntBitOr(lhsPtr, rhsPtr, nLimbs, orPtr)
console_log(mload(orPtr)) // 0xff
console_log(mload(add(orPtr, 0x20))) // 0xff

// 0101 0101 | 1010 1010 = 1111 1111 (needed?)
mstore(lhsPtr, 0x5555555555555555555555555555555555555555555555555555555555555555)
mstore(add(lhsPtr, 0x20), 0x5555555555555555555555555555555555555555555555555555555555555555)
mstore(rhsPtr, 0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa)
mstore(add(rhsPtr, 0x20), 0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa)
bigUIntBitOr(lhsPtr, rhsPtr, nLimbs, orPtr)
console_log(mload(orPtr)) // 0xff
console_log(mload(add(orPtr, 0x20))) // 0xff

// 0000 1111 | 0000 1111 = 0000 1111
mstore(lhsPtr, 0x00)
mstore(add(lhsPtr, 0x20), 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
mstore(rhsPtr, 0x00)
mstore(add(rhsPtr, 0x20), 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
bigUIntBitOr(lhsPtr, rhsPtr, nLimbs, orPtr)
console_log(mload(orPtr)) // 0x00
console_log(mload(add(orPtr, 0x20))) // 0xff

// 1111 0000 | 1111 0000 = 1111 0000
mstore(lhsPtr, 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
mstore(add(lhsPtr, 0x20), 0x00)
mstore(rhsPtr, 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
mstore(add(rhsPtr, 0x20), 0x00)
bigUIntBitOr(lhsPtr, rhsPtr, nLimbs, orPtr)
console_log(mload(orPtr)) // 0xff
console_log(mload(add(orPtr, 0x20))) // 0x00

// 0000 1111 | 1111 0000 = 1111 1111
mstore(lhsPtr, 0x00)
mstore(add(lhsPtr, 0x20), 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
mstore(rhsPtr, 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
mstore(add(rhsPtr, 0x20), 0x00)
bigUIntBitOr(lhsPtr, rhsPtr, nLimbs, orPtr)
console_log(mload(orPtr)) // 0xff
console_log(mload(add(orPtr, 0x20))) // 0xff
```
3. Run the node.
4. Make the following call to the node:
```shell
zksync-era-cli --host localhost --port 8011 call --contract 0x000000000000000000000000000000000000FFFF --function "" --data 00  --private-key 0x850683b40d4a740aa6e745f889a6fdc8327be76e122f5aba645a5b02d0248db8
```
5. You should see the following output in the node:
```shell
Call SUCCESS
=== Console Logs:
0
ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
0
0
ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
0
ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
0
ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
```

**Pull request merged**